src/Home.tsx

```diff
@@ -102,55 +113,78 @@ const Home = ({
 			if (retryIndex === null) {
 				await writableStream?.write(new Int32Array([blockCount]).buffer);
 			} else {
-				const offset: any = (await fileHandle?.getFile())?.size;
-				writableStream?.seek(offset)
+				const offset: number = (await fileHandle?.getFile())?.size ?? 0;
+      	await writableStream?.seek(offset);
 			}
 
-			const indexFileStart = Math.floor(formData.spanStart / 128000);
-			const indexFileCount = Math.ceil((formData.spanEnd - formData.spanStart) / 128000) + indexFileStart;
-			for (let indexFile = indexFileStart; indexFile <= indexFileCount; indexFile += 1) {
-				if (controllerStop.signal.aborted) throw new Error('Fetching aborted');
-				if (controllerPause.signal.aborted) throw new Error('Paused');
+			const queue: QueuedBlock[] = [];
+			const startBlock: any = retryIndex !== null ? retryIndex : formData.spanStart;
+			let lastBlockIndex: number | null = null;
+			let writtenCount: number = retryIndex !== null ? retryIndex : 0;
+			let isRefilling = false;
 
-				const indexData: Uint8Array | string = await fetchIndexFile(currentNet, indexFile);
-				if (typeof indexData === 'string') {
-					await writableStream?.close();
-					onModal('failed', indexData, (retryIndexTemp: number) => fetchBlocksInRange(+formData.spanStart + retryIndexTemp));
-					return
+			const refillQueue = async (nextBlock: number = startBlock) => {
+				if (controllerStop.signal.aborted) return;
+				if (isRefilling || nextBlock > +formData.spanEnd) return;
+    		isRefilling = true;
+
+				const remaining = +formData.spanEnd - nextBlock + 1;
+				const limit = Math.min(1000, remaining);
+				try {
+					const objectsData: BlockWithAttrs[] | string = await fetchBatchBlocks(currentNet, nextBlock, limit);
+					if (typeof objectsData === 'string') {
+						controllerStop.abort();
+						await writableStream?.close();
+						onModal('failed', objectsData, (retryIndexTemp: number) => fetchBlocksInRange(nextBlock + retryIndexTemp));
+						return
+					}
+
+					objectsData.forEach((block) => queue.push({ blockNumber: Number(block.attributes.Block), oid: block.objectId }));
+				} finally {
+					isRefilling = false;
 				}
+			};
+
+			await refillQueue();
 
-				const uint8Data = new Uint8Array(indexData);
-				const objectsData: string[] = [];
-				for (let i = 0; i < uint8Data.length; i += 32) {
-					const chunk = uint8Data.slice(i, i + 32);
-					const encoded = base58.encode(chunk);
-					objectsData.push(encoded);
+			while (!controllerStop.signal.aborted && writtenCount < blockCount) {
+				if (queue.length < 500 && queue.length != 0) {
```

Comment:
Why `!= 0`? Don't we want to continue if queue is depleted?

Reply:
Yes indeed I did, it's a useless condition